localstack/services/sns/provider.py

```diff
@@ -155,6 +162,11 @@ def publish_message(
         # TODO: check this in the old provider
         cache.append(req_data)
 
+        message_structure = req_data.get("MessageStructure", [None])[0]
+        LOG.debug("Publishing message to Endpoint: %s | Message: %s", target_arn, message)
+        message_to_endpoint(target_arn, message, message_structure)
+        return message_id
```

Comment:
See my general comment: we could only have the part sending the request be in a thread, but all the checks about the endpoint/platform be before that, to be able to raise exceptions if the call is wrong (wrong endpoint).

Reply:
publishing to a target directly can raise an exception depending on the platform response. now the validation of the endpoint is done before the call of that function but it still needs to be sync.